test/replication-luatest/gh_6842_qsync_applier_order_test.lua

```diff
@@ -0,0 +1,370 @@
+        f1:set_joinable(true)
+        local f2 = fiber.new(s.replace, s, {2})
+        f2:set_joinable(true)
+        return {f1:id(), f2:id()}
+    end)
+
+    -- The txns are delivered to server 2.
+    wait_synchro_len(g.server2, #fids)
+
+    -- Server 2 confirms the txns and sends a PROMOTE covering it to server 1.
+    g.server2:exec(function()
+        box.cfg{
+            replication_synchro_quorum = 1,
+            -- To make it not wait for confirm from server 1. Just confirm via
+            -- the promotion ASAP.
+            replication_synchro_timeout = 0.001,
+        }
+        box.ctl.promote()
+        box.cfg{
+            replication_synchro_quorum = 2,
+            replication_synchro_timeout = 1000,
+        }
+    end)
+
+    -- Server 1 receives the foreign PROMOTE. Now the local txns and a remote
+    -- PROMOTE are being written to WAL.
+    wait_synchro_is_busy(g.server1)
+
+    local rows = g.server1:exec(function(fids)
+        -- Simulate the TX thread being slow. To increase likelihood of the txn
+        -- and PROMOTE WAL writes being processed by TX in a single batch.
+        box.error.injection.set('ERRINJ_TX_DELAY_PRIO_ENDPOINT', 0.1)
+        -- Let them finish WAL writes.
+        box.error.injection.set('ERRINJ_RELAY_FASTER_THAN_TX', false)
+
+        local fiber = require('fiber')
+        for _, fid in pairs(fids) do
+            fiber.find(fid):join()
+        end
+        box.error.injection.set('ERRINJ_TX_DELAY_PRIO_ENDPOINT', 0)
+        return box.space.test:select()
+    end, {fids})
+    -- The txn was confirmed by another instance even before it was confirmed
+    -- locally.
+    luatest.assert_equals(rows, {{1}, {2}}, 'txn was confirmed')
+
+    local rows2 = g.server2:exec(function()
+        return box.space.test:select()
+    end)
+    luatest.assert_equals(rows2, rows, 'on instance 2 too')
+
+    -- The synchronous replication is functional - new owner can use the queue.
+    g.server2:exec(function()
+        box.space.test:replace{3}
+    end)
+    g.server1:wait_vclock_of(g.server2)
+    rows = g.server1:exec(function()
+        return box.space.test:select{}
+    end)
+    luatest.assert_equals(rows, {{1}, {2}, {3}}, 'synchro transactions work')
+end
+
+--
+-- Server 1 was a synchro queue owner. It starts a synchro txn. The txn is
+-- not written to WAL yet. Server 2 makes a PROMOTE which doesn't include that
+-- txn.
+--
+-- Then the PROMOTE is delivered to server 1 and goes to WAL too. Then the txn
+-- and PROMOTE WAL writes are processed by TX thread in a single batch without
+-- yields.
+--
+-- The bug was that the fiber authored the txn wasn't ready to a WAL write
+-- being success, txn being marked as non-synchro anymore (it happened on
+-- rollback in the synchro queue), but the txn signature being bad.
+--
+g.test_remote_promote_during_local_txn_not_including_it = function(g)
+    -- Start a synchro txn on server 1.
+    local fids = g.server1:exec(function()
```

Comment:
Let's extract that `exec` into a helper? 
It repeats in 2 tests and is quite large.

Reply:
Not precisely repeated. In `test_local_txn_during_remote_promote` I don't do DML and use `ERRINJ_WAL_DELAY_COUNTDOWN`. In `test_remote_promote_during_local_txn_including_it` I use `ERRINJ_RELAY_FASTER_THAN_TX`. In `test_remote_promote_during_local_txn_not_including_it` I use `ERRINJ_WAL_DELAY`. We could in theory try to extract common parts of the last two, but I suspect it will only make the tests harder to read - it is simpler when text-specific context is close except for very common parts like `wait_synchro_is_busy`. WDYT?

Reply:
I"d probably make the errinj a function argument. Other than the errinj `exec`s in `test_remote_promote_during_local_txn_including_it` and `test_remote_promote_during_local_txn_not_including_it` are the same.
I don't insist, though. Maybe the tests are indeed easier to read now.

Reply:
Ok, I tried now. Stumbled upon issues with readability indeed.

Lets try to extract the entire function body with only errinj name parametrized. Then firstly, we have an issue with name. It would be something as `promote_and_hang_fiber_txns(server, tuples, errinj)`. It looks not too bad, but later the space created inside of the function is going to be used in the test openly in its end where I test `-- The synchronous replication is functional` (not in the other big exec which could also be extracted).

So the second issue is that there are things done inside of the exec which were not parametrized and we rely on them. Then lets try to keep space creation out of this new function. In that case we also need to keep `box.ctl.promote()` out of it - it is done before space creation. Now we have this code left:
```Lua
        box.cfg{
            -- To hang own transactions in the synchro queue.
            replication_synchro_quorum = 3,
        }
        box.error.injection.set('ERRINJ_RELAY_FASTER_THAN_TX', true)
        local fiber = require('fiber')
        -- More than one transaction to ensure that it works not just for one.
        local f1 = fiber.new(s.replace, s, {1})
        f1:set_joinable(true)
        local f2 = fiber.new(s.replace, s, {2})
        f2:set_joinable(true)
        return {f1:id(), f2:id()}
```
Now it operates on a space created outside. It means the function gets a new parameter - space name: `hang_txns(server, tuples, space, errinj)`. Note, we need to pass tuples too, because later I rely on tuple `{3}` not being inserted yet. I rely on the space content out of that function.

The next issue is that I set the quorum to 3 here, but the function name didn't tell that. So if afterwards on the same server in a test case I will do something quorum-related, it will break and from the test case code it won't be clear why.

Here I stopped. The best solution would be to extract the `fiber.new()` + `fiber:set_joinable()` and then `fiber:join()` loop. But these will make more `:exec()` calls leading to ever bulkier code.

Please, lets keep it as is?